srk/src/nestedLoops.ml

```diff
@@ -0,0 +1,311 @@
+    begin
+        printf "this loop is a terminal loop";
+        let (hd_pred, hd, be) = loop.back_edge in
+        let bf = compute_body_with_header_and_backedge wg_no_bg hd hd_pred be in
+        let (wgnn, splitted_vertex) = split_header wg_no_bg hd bf loop in
+            (wgnn,
+            { loop with splitted_hd = splitted_vertex; body_f = bf}) 
+    end
+    else 
+        (* simultaneously compute transformed graph and list of processed children *)
+        let (wgg, processed_children) = 
+            List.fold_right
+                (fun child tup -> 
+                    let (wgt, list) = tup in
+                    let (wgtt, processed_loop) = compute_body_weight wgt child in
+                    (wgtt, processed_loop :: list)
+                )
+                loop.children
+                (wg_no_bg, [])
+        in
+        let finished_children = processed_children in
+        let (hd_pred, hd, be) = loop.back_edge in
+        let bf = compute_body_with_header_and_backedge wgg hd hd_pred be in
+            let (wgnn, splitted_vertex) = split_header wgg hd bf loop in
+            (wgnn, 
+            { loop with splitted_hd = splitted_vertex; body_f = bf; children = finished_children})
+
+
+(** Compute a path from beginning to the header of this loop and all its children loops,
+    given a weighted graph with all back edges removed and header splitted to contain
+    the label of the loop body. 
+    `start_to_v` is a function that computes the path from the entry to any vertex. 
+    `combine` is a function that constructs the path in the form header * star(body). 
+ *)
+let rec compute_path_to_header wg_no_bg_and_splitted_headers loop start =
+    let construct_header_guard path_to_hd body_f =
+        let alg = get_algebra wg_no_bg_and_splitted_headers in
+        alg.mul path_to_hd (alg.star body_f)
+    in
+    let splitted_hd = loop.splitted_hd in
+    if splitted_hd = loop.header then
+        assert false
+    else
+        let bf = loop.body_f in
+        if loop.children = [] then
+            {
+                loop with 
+                    header_f = construct_header_guard 
+                                    (path_weight wg_no_bg_and_splitted_headers start splitted_hd) 
+                                    bf
+            }
+        else
+            let processed_children =
+                List.map
+                    (fun child ->
+                        compute_path_to_header 
+                            wg_no_bg_and_splitted_headers  
+                            child  
+                            start)
+                    loop.children
+            in
+            {
+                loop with
+                    header_f = construct_header_guard 
+                                    (path_weight wg_no_bg_and_splitted_headers start splitted_hd) 
+                                    bf;
+                    children = processed_children
+            }
+
+let rec flatten_loops loops_l results =
```

Comment:
This is equivalent to
List.rev_map (fun loop -> (loop.header_f, loop.body_f)) loops_l

I think "loops_l" is intended to expand to "loops_list" --- in OCaml, the "s" conventionally designates a list (loops is a list of things, each of which is a loop).